pandas/core/ops/__init__.py

```diff
@@ -722,13 +722,28 @@ def _arith_method_FRAME(cls, op, special):
     @Appender(doc)
     def f(self, other, axis=default_axis, level=None, fill_value=None):
 
+        orig_self, orig_other = self, other
         self, other = _align_method_FRAME(self, other, axis, flex=True, level=level)
 
         if isinstance(other, ABCDataFrame):
             # Another DataFrame
             pass_op = op if should_series_dispatch(self, other, op) else na_op
             pass_op = pass_op if not is_logical else op
 
+            if isinstance(orig_other, ABCDataFrame):
+                if fill_value is None and level is None and axis is default_axis:
+                    # TODO: any other cases we should handle here?
+                    cols = orig_self.columns.intersection(orig_other.columns)
+                    if not (
+                        cols.equals(orig_self.columns)
+                        and cols.equals(orig_other.columns)
+                    ):
+                        # GH#31623, only operate on shared columns
+                        new_left = orig_self[cols]
+                        new_right = orig_other[cols]
+                        result = op(new_left, new_right)
+                        return result.reindex(self.columns, axis=1)
```

Comment:
This should be the union of the columns right?

Reply:
i think so, but id have to double-check the exact outer-join semantics